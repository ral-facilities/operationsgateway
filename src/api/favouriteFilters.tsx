import {
  useMutation,
  UseMutationResult,
  useQuery,
  useQueryClient,
  UseQueryResult,
} from '@tanstack/react-query';
import axios, { AxiosError } from 'axios';
import { FavouriteFilter, FavouriteFilterPost } from '../app.types';
import { readSciGatewayToken } from '../parseTokens';
import { useAppSelector } from '../state/hooks';
import { selectUrls } from '../state/slices/configSlice';

const addFavouriteFilter = (
  apiUrl: string,
  favouriteFilter: FavouriteFilterPost
): Promise<string> => {
  const queryParams = new URLSearchParams();

  queryParams.append('name', favouriteFilter.name);
  queryParams.append('filter', favouriteFilter.filter);

  return axios
    .post<string>(
      `${apiUrl}/users/filters`,
      {},
      {
        params: queryParams,
        headers: {
          Authorization: `Bearer ${readSciGatewayToken()}`,
        },
      }
    )
    .then((response) => response.data);
};

export const useAddFavouriteFilter = (): UseMutationResult<
  string,
  AxiosError,
  FavouriteFilterPost
> => {
  const { apiUrl } = useAppSelector(selectUrls);
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (favouriteFilter: FavouriteFilterPost) =>
      addFavouriteFilter(apiUrl, favouriteFilter),
    onError: (error) => {
      console.log('Got error ' + error.message);
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['favouriteFilters'] });
    },
  });
};

const fetchFavouriteFilters = (apiUrl: string): Promise<FavouriteFilter[]> => {
  return axios
    .get(`${apiUrl}/users/filters`, {
      headers: {
        Authorization: `Bearer ${readSciGatewayToken()}`,
      },
    })
    .then((response) => {
      return response.data;
    });
};

export const useFavouriteFilters = (): UseQueryResult<
  FavouriteFilter[],
  AxiosError
> => {
  const { apiUrl } = useAppSelector(selectUrls);

  return useQuery({
    queryKey: ['favouriteFilters'],

    queryFn: () => {
      return fetchFavouriteFilters(apiUrl);
    },
  });
};
