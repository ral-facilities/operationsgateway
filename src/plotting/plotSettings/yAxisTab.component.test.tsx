import type { RenderResult } from '@testing-library/react';
import { fireEvent, render, screen, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { staticChannels } from '../../api/channels';
import { FullScalarChannelMetadata } from '../../app.types';
import { testScalarChannels } from '../../testUtils';
import { COLOUR_ORDER } from './colourGenerator';
import type { YAxisTabProps } from './yAxisTab.component';
import YAxisTab from './yAxisTab.component';

describe('y-axis tab', () => {
  let props: YAxisTabProps;
  let user: ReturnType<typeof userEvent.setup>;
  const changeLeftYAxisScale = vi.fn();
  const changeRightYAxisScale = vi.fn();
  const changeSelectedPlotChannels = vi.fn();
  const changeLeftYAxisMinimum = vi.fn();
  const changeLeftYAxisMaximum = vi.fn();
  const changeRightYAxisMinimum = vi.fn();
  const changeRightYAxisMaximum = vi.fn();
  const changeSelectedColours = vi.fn();
  const changeRemainingColours = vi.fn();
  const changeRightYAxisLabel = vi.fn();
  const changeLeftYAxisLabel = vi.fn();

  const createView = (): RenderResult => {
    return render(<YAxisTab {...props} />);
  };

  beforeEach(() => {
    props = {
      selectedRecordTableChannels: [
        staticChannels['timestamp'] as FullScalarChannelMetadata,
      ],
      allChannels: testScalarChannels,
      selectedPlotChannels: [],
      changeSelectedPlotChannels,
      changeLeftYAxisMinimum,
      changeLeftYAxisMaximum,
      changeRightYAxisMinimum,
      changeRightYAxisMaximum,
      leftYAxisScale: 'linear',
      rightYAxisScale: 'logarithmic',
      changeLeftYAxisScale,
      changeRightYAxisScale,
      initialSelectedColours: [],
      initialRemainingColours: COLOUR_ORDER.map((colour) => colour),
      changeSelectedColours,
      changeRemainingColours,
      changeRightYAxisLabel,
      changeLeftYAxisLabel,
    };

    user = userEvent.setup({ delay: null });
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  it('renders correctly', () => {
    const { asFragment } = createView();

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly with selected channels', () => {
    props.selectedPlotChannels = testScalarChannels.map((channel) => ({
      name: channel.systemName,
      units: 'mm',
      displayName: channel.name,
      options: {
        visible: true,
        colour: '#ffffff',
        lineStyle: 'solid',
        yAxis: 'left',
      },
    }));

    createView();

    props.selectedPlotChannels.forEach((channel) => {
      const channelName = channel.displayName ?? channel.name;
      const channelLabel = screen.getByLabelText(`${channelName} label`);
      expect(within(channelLabel).getByText(channelName)).toBeInTheDocument();
      expect(
        within(channelLabel).getByLabelText(`More options for ${channelName}`)
      ).toBeInTheDocument();
      expect(
        within(channelLabel).getByLabelText(`Remove ${channelName} from plot`)
      ).toBeInTheDocument();
    });
  });

  it('renders Y scale radio buttons and calls changeLeftYAxisScale on click when the axis selected is the left', async () => {
    createView();

    const radioGroup = screen.getByRole('radiogroup', { name: 'Scale' });
    expect(
      within(radioGroup).getByRole('radio', {
        name: 'Linear',
      })
    ).toBeChecked();

    await user.click(screen.getByRole('radio', { name: 'Log' }));

    expect(changeLeftYAxisScale).toHaveBeenCalledWith('logarithmic');
  });

  it('renders Y scale radio buttons and calls changeRightYAxisScale on click when the axis selected is the right', async () => {
    createView();

    await user.click(screen.getByRole('button', { name: 'Right' }));

    const radioGroup = screen.getByRole('radiogroup', { name: 'Scale' });
    expect(
      within(radioGroup).getByRole('radio', {
        name: 'Log',
      })
    ).toBeChecked();

    await user.click(screen.getByRole('radio', { name: 'Linear' }));

    expect(changeRightYAxisScale).toHaveBeenCalledWith('linear');
  });

  it('allows user to add channels on the y-axis (keyboard only)', async () => {
    createView();

    const autocomplete = screen.getByTestId('data-channels-autocomplete');
    const input = within(autocomplete).getByRole('combobox');

    await user.type(input, 'Channel_');
    autocomplete.focus();
    fireEvent.keyDown(autocomplete, { key: 'ArrowDown' });
    fireEvent.keyDown(autocomplete, { key: 'Enter' });

    expect(changeSelectedPlotChannels).toHaveBeenCalledWith([
      {
        displayName: 'Channel_ABCDE',
        name: 'CHANNEL_ABCDE',
        units: '',
        options: {
          visible: true,
          colour: expect.anything(),
          lineStyle: 'solid',
          yAxis: 'left',
        },
      },
    ]);
  });

  it('allows user to add channels on the y-axis (mouse and keyboard)', async () => {
    createView();

    await user.click(screen.getByRole('button', { name: 'Right' }));

    const autocomplete = screen.getByTestId('data-channels-autocomplete');
    const input = within(autocomplete).getByRole('combobox');

    await user.type(input, 'Channel_');
    await user.click(screen.getByText('Channel_DEFGH'));

    expect(changeSelectedPlotChannels).toHaveBeenCalledWith([
      {
        displayName: 'Channel_DEFGH',
        name: 'CHANNEL_DEFGH',
        units: '',
        options: {
          visible: true,
          colour: expect.anything(),
          lineStyle: 'solid',
          yAxis: 'right',
        },
      },
    ]);
  });

  it('populates the displayed table channels dropdown and adds selection to the y-axis', async () => {
    props.selectedRecordTableChannels = testScalarChannels.filter(
      (channel) => channel.systemName === 'CHANNEL_ABCDE'
    );
    createView();

    const select = screen.getByLabelText('Displayed table channels');
    await userEvent.click(select);

    const dropdown = screen.getByRole('listbox', {
      name: 'Displayed table channels',
    });
    await userEvent.click(
      within(dropdown).getByRole('option', { name: 'Channel_ABCDE' })
    );

    expect(changeSelectedPlotChannels).toHaveBeenCalledWith([
      {
        name: 'CHANNEL_ABCDE',
        displayName: 'Channel_ABCDE',
        options: {
          visible: true,
          colour: expect.anything(),
          lineStyle: 'solid',
          yAxis: 'left',
        },
      },
    ]);
  });

  it('only populates the displayed table channels dropdown with options not already selected', async () => {
    props.selectedRecordTableChannels = testScalarChannels.filter(
      (channel) => channel.systemName === 'CHANNEL_ABCDE'
    );
    props.selectedPlotChannels = [
      {
        name: 'CHANNEL_ABCDE',
        units: 'mm',
        displayName: 'Channel_ABCDE',
        options: {
          visible: true,
          colour: '#ffffff',
          lineStyle: 'solid',
          yAxis: 'left',
        },
      },
    ];
    createView();

    const select = screen.getByLabelText('Displayed table channels');
    await userEvent.click(select);

    const dropdown = screen.getByRole('listbox', {
      name: 'Displayed table channels',
    });
    expect(within(dropdown).queryByRole('option')).not.toBeInTheDocument();
  });

  it('removes channel from display when we click Close on its label', async () => {
    props.selectedPlotChannels = [
      {
        name: 'CHANNEL_ABCDE',
        units: 'mm',
        displayName: 'Channel_ABCDE',
        options: {
          visible: true,
          colour: '#ffffff',
          lineStyle: 'solid',
          yAxis: 'left',
        },
      },
      {
        name: 'CHANNEL_DEFGH',
        units: 'cm',
        displayName: 'Channel_DEFGH',
        options: {
          visible: true,
          colour: '#ffffff',
          lineStyle: 'solid',
          yAxis: 'left',
        },
      },
    ];
    createView();

    await user.click(screen.getByLabelText('Remove Channel_ABCDE from plot'));
    expect(changeSelectedPlotChannels).toHaveBeenLastCalledWith([
      {
        name: 'CHANNEL_DEFGH',
        units: 'cm',
        displayName: 'Channel_DEFGH',
        options: {
          visible: true,
          colour: '#ffffff',
          lineStyle: 'solid',
          yAxis: 'left',
        },
      },
    ]);
    expect(changeLeftYAxisScale).not.toHaveBeenCalled();
    expect(changeRightYAxisScale).not.toHaveBeenCalled();
  });

  it('removes channel from display when we click Close on its label and resets y-axis scale to linear if no selected channels remain', async () => {
    props.selectedPlotChannels = [
      {
        name: 'CHANNEL_ABCDE',
        units: 'mm',
        displayName: 'Channel_ABCDE',
        options: {
          visible: true,
          colour: '#ffffff',
          lineStyle: 'solid',
          yAxis: 'left',
        },
      },
    ];
    createView();

    await user.click(screen.getByLabelText('Remove Channel_ABCDE from plot'));
    expect(changeSelectedPlotChannels).toHaveBeenLastCalledWith([]);
    expect(changeLeftYAxisScale).toHaveBeenCalledWith('linear');
    expect(changeRightYAxisScale).toHaveBeenCalledWith('linear');
  });

  describe('min and max fields', () => {
    it('lets user change the min field and calls relevant onchange method', async () => {
      createView();

      const minField = screen.getByLabelText('Min');
      await user.type(minField, '0');
      expect(changeLeftYAxisMinimum).toHaveBeenCalledWith(0);
    });

    it('lets user change the max field and calls relevant onchange method', async () => {
      createView();

      await user.click(screen.getByRole('button', { name: 'Right' }));

      const maxField = screen.getByLabelText('Max');
      await user.type(maxField, '0');
      expect(changeRightYAxisMaximum).toHaveBeenCalledWith(0);
    });

    it('sets minimum value to undefined if no float value is present', async () => {
      createView();

      await user.click(screen.getByRole('button', { name: 'Right' }));

      const minField = screen.getByLabelText('Min');
      await user.type(minField, '1');
      expect(changeRightYAxisMinimum).toHaveBeenLastCalledWith(1);

      await user.clear(minField);
      expect(changeRightYAxisMinimum).toHaveBeenLastCalledWith(undefined);
    });

    it('sets maximum value to undefined if no float value is present', async () => {
      createView();

      const maxField = screen.getByLabelText('Max');
      await user.type(maxField, '1');
      expect(changeLeftYAxisMaximum).toHaveBeenLastCalledWith(1);

      await user.clear(maxField);
      expect(changeLeftYAxisMaximum).toHaveBeenLastCalledWith(undefined);
    });

    it('displays helper text when min and max fields contain an invalid range', async () => {
      createView();

      const minField = screen.getByLabelText('Min');
      const maxField = screen.getByLabelText('Max');
      await user.type(minField, '2');
      await user.type(maxField, '1');

      // Check the helper text displays
      screen.getAllByText('Invalid range');

      // One for each input box
      expect(screen.getAllByText('Invalid range').length).toEqual(2);
    });

    it('initialises correctly with falsy values', async () => {
      props.initialLeftYAxisMinimum = 0;
      props.initialLeftYAxisMaximum = 0;
      createView();

      const minField = screen.getByLabelText('Min');
      expect(minField).toHaveValue('0');

      const maxField = screen.getByLabelText('Max');
      expect(maxField).toHaveValue('0');
    });
  });

  it('allows customization of left y-axis label', async () => {
    function TestEnv() {
      const [leftYAxisLabel, setLeftYAxisLabel] = React.useState('');
      return (
        <YAxisTab
          {...props}
          leftYAxisLabel={leftYAxisLabel}
          changeLeftYAxisLabel={setLeftYAxisLabel}
        />
      );
    }

    render(<TestEnv />);

    const axisLabelTextBox = screen.getByRole('textbox', { name: 'Label' });
    expect(axisLabelTextBox).toHaveValue('');

    await user.type(axisLabelTextBox, 'Left axis');
    expect(axisLabelTextBox).toHaveValue('Left axis');
  });

  it('allows customization of right y-axis label', async () => {
    function TestEnv() {
      const [rightYAxisLabel, setRightYAxisLabel] = React.useState('');
      return (
        <YAxisTab
          {...props}
          leftYAxisLabel={rightYAxisLabel}
          changeLeftYAxisLabel={setRightYAxisLabel}
        />
      );
    }

    render(<TestEnv />);

    const axisLabelTextBox = screen.getByRole('textbox', { name: 'Label' });
    expect(axisLabelTextBox).toHaveValue('');

    await user.type(axisLabelTextBox, 'Right axis');
    expect(axisLabelTextBox).toHaveValue('Right axis');
  });
});
