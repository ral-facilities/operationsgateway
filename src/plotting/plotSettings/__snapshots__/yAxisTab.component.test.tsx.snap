// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`y-axis tab > renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 css-1w7ngrt-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <div
        aria-label="y axis toggle"
        class="MuiToggleButtonGroup-root css-bghv3y-MuiToggleButtonGroup-root"
        role="group"
      >
        <button
          aria-pressed="true"
          class="MuiButtonBase-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButton-root Mui-selected MuiToggleButton-sizeSmall MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButtonGroup-firstButton css-1d80cto-MuiButtonBase-root-MuiToggleButton-root"
          tabindex="0"
          type="button"
          value="left"
        >
          Left
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          aria-pressed="false"
          class="MuiButtonBase-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButton-root MuiToggleButton-sizeSmall MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButtonGroup-lastButton css-1d80cto-MuiButtonBase-root-MuiToggleButton-root"
          tabindex="0"
          type="button"
          value="right"
        >
          Right
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item css-1f064cs-MuiGrid-root"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":r0:"
          id=":r0:-label"
          style="font-size: 12px;"
        >
          Label
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          style="font-size: 12px;"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r0:"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Label
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-spacing-xs-1 css-v8g4am-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":r1:"
            id=":r1:-label"
            style="font-size: 12px;"
          >
            Min
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            style="font-size: 12px;"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r1:"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Min
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":r2:"
            id=":r2:-label"
            style="font-size: 12px;"
          >
            Max
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            style="font-size: 12px;"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r2:"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Max
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-wrap-xs-nowrap css-at2ojq-MuiGrid-root"
    >
      <div
        class="MuiFormControl-root css-qtsj9i-MuiFormControl-root"
      >
        <label
          class="MuiFormLabel-root MuiFormLabel-colorPrimary css-7dax6m-MuiFormLabel-root"
          id="left-y-scale-group-label"
        >
          Scale
        </label>
        <div
          aria-labelledby="left-y-scale-group-label"
          class="MuiFormGroup-root MuiFormGroup-row MuiRadioGroup-root MuiRadioGroup-row css-qfz70r-MuiFormGroup-root"
          role="radiogroup"
        >
          <label
            class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
          >
            <span
              class="MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root MuiRadio-root MuiRadio-colorPrimary Mui-checked MuiRadio-root MuiRadio-colorPrimary css-vqmohf-MuiButtonBase-root-MuiRadio-root"
            >
              <input
                checked=""
                class="PrivateSwitchBase-input css-1m9pwf3"
                name="left y scale radio buttons group"
                type="radio"
                value="linear"
              />
              <span
                class="css-hyxlzm"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hbvpl3-MuiSvgIcon-root"
                  data-testid="RadioButtonUncheckedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-11zohuh-MuiSvgIcon-root"
                  data-testid="RadioButtonCheckedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
            >
              Linear
            </span>
          </label>
          <label
            class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
          >
            <span
              class="MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root MuiRadio-root MuiRadio-colorPrimary MuiRadio-root MuiRadio-colorPrimary css-vqmohf-MuiButtonBase-root-MuiRadio-root"
            >
              <input
                class="PrivateSwitchBase-input css-1m9pwf3"
                name="left y scale radio buttons group"
                type="radio"
                value="logarithmic"
              />
              <span
                class="css-hyxlzm"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hbvpl3-MuiSvgIcon-root"
                  data-testid="RadioButtonUncheckedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hhw7if-MuiSvgIcon-root"
                  data-testid="RadioButtonCheckedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
            >
              Log
            </span>
          </label>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item css-1f064cs-MuiGrid-root"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1pddimr-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          id="table-channel-select-label"
        >
          Displayed table channels
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-i06ziq-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
        >
          <div
            aria-controls=":r4:"
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby="table-channel-select-label"
            class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
            role="combobox"
            tabindex="0"
          >
            <span
              class="notranslate"
            >
              ​
            </span>
          </div>
          <input
            aria-hidden="true"
            aria-invalid="false"
            class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
            tabindex="-1"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Displayed table channels
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item css-1f064cs-MuiGrid-root"
    >
      <div
        class="MuiAutocomplete-root MuiAutocomplete-fullWidth MuiAutocomplete-hasPopupIcon css-fzem9i-MuiAutocomplete-root"
        data-testid="data-channels-autocomplete"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="select data channels"
            id="select data channels-label"
            style="font-size: 12px;"
          >
            Search all channels
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedStart MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-1gywuxd-MuiInputBase-root-MuiOutlinedInput-root"
            style="font-size: 12px;"
          >
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeSmall css-ittuaa-MuiInputAdornment-root"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="SearchIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
                />
              </svg>
            </div>
            <input
              aria-autocomplete="list"
              aria-expanded="false"
              aria-invalid="false"
              autocapitalize="none"
              autocomplete="off"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-12yjm75-MuiInputBase-input-MuiOutlinedInput-input"
              id="select data channels"
              role="combobox"
              spellcheck="false"
              type="text"
              value=""
            />
            <div
              class="MuiAutocomplete-endAdornment css-p1olib-MuiAutocomplete-endAdornment"
            >
              <button
                aria-label="Open"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                tabindex="-1"
                title="Open"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-14lo706"
              >
                <span>
                  Search all channels
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
